grove/arbitrary_state/arbitrary_state.py

```diff
@@ -0,0 +1,214 @@
+                             of vector to be present as a coefficient in the
+                             final state. If more than :math:`n` are provided,
+                             only the first :math:`n` will be used.
+                             If no list is provided, the default will be
+                             qubits :math:`0, 1, \ldots, n-1`.
+    :return: a program that takes in :math:`\\vert 0\\rangle^{\otimes n}`
+             and produces a state that represents this vector, as
+             described above.
+    :rtype: Program
+    """
+    vec_norm = vector / np.linalg.norm(vector)
+    n = max(1, int(np.ceil(np.log2(len(vec_norm)))))  # number of qubits needed
+
+    if qubits is None:
+        qubits = range(n)
+
+    N = 2 ** n  # number of coefficients
+    while len(vec_norm) < N:
+        vec_norm = np.append(vec_norm, 0)  # pad with zeros
+
+    magnitudes = map(np.abs, vec_norm)
+    phases = map(np.angle, vec_norm)
+
+    # because conceptually this algorithm starts with the end state
+    # and makes it into the |0> state,
+    # the gates will be constructed in reverse order
+    reversed_gates = []
+
+    # matrix that converts angles of uniformly controlled rotation
+    # to angles of uncontrolled rotations
+    # at first, all qubits except for the 0 qubit act as controls
+    M = get_uniformly_controlled_rotation_matrix(n - 1)
+
+    # generate the positions of the controls for the CNOTs
+    # at first, all qubits except for the 0 qubit act as controls
+    rotation_cnots = get_cnot_control_positions(n - 1)
+
+    for step in xrange(n):
+        # will hold the angles for controlled rotations
+        # in the phase unification and probability unification steps,
+        # respectively
+        z_thetas = []
+        y_thetas = []
+
+        # will hold updated phases and magnitudes after rotations
+        new_phases = []
+        new_magnitudes = []
+
+        for i in xrange(0, len(phases), 2):
+            # find z rotation angles
+            phi = phases[i]
+            psi = phases[i + 1]
+            z_thetas.append(phi - psi)
+
+            # update phases after applying such rotations
+            kappa = (phi + psi) / 2.
+            new_phases.append(kappa)
+
+            # find y rotation angles
+            a = magnitudes[i]
+            b = magnitudes[i + 1]
+            if a == 0 and b == 0:
+                y_thetas.append(0)
+            else:
+                y_thetas.append(
+                    2 * np.arcsin((a - b) / (np.sqrt(2 * (a ** 2 + b ** 2)))))
+
+            # update magnitudes after applying such rotations
+            c = np.sqrt((a ** 2 + b ** 2) / 2.)
+            new_magnitudes.append(c)
+
+        # convert these rotation angles
+        # to those for uncontrolled rotations + CNOTs
+        converted_z_thetas = np.dot(M, z_thetas)
+        converted_y_thetas = np.dot(M, y_thetas)
+
+        # phase unification
+        for j in xrange(len(converted_z_thetas)):
```

Comment:
Make a separate function so you can check correctness of the programs coming out of each phase unification and probability unification step.